alab_management/resource_manager/resource_manager.py

```diff
@@ -0,0 +1,207 @@
+        self.sample_view = SampleView()
+        self.device_view = DeviceView()
+        self._request_collection = get_collection("requests")
+
+        self.logger = DBLogger(task_id=None)
+        super().__init__()
+        time.sleep(1)  # allow some time for other modules to launch
+
+    def run(self):
+        """Start the loop."""
+        while True:
+            self._loop()
+            # time.sleep(1)
+
+    def _loop(self):
+        self.handle_released_resources()
+        self.handle_requested_resources()
+
+    def handle_released_resources(self):
+        """Release the resources."""
+        for request in self.get_requests_by_status(RequestStatus.NEED_RELEASE):
+            devices = request["assigned_devices"]
+            sample_positions = request["assigned_sample_positions"]
+            self._release_devices(devices)
+            self._release_sample_positions(sample_positions)
+            self.update_request_status(
+                request_id=request["_id"], status=RequestStatus.RELEASED
+            )
+
+    def handle_requested_resources(self):
+        """
+        Check if there are any requests that are in PENDING status. If so,
+        try to assign the resources to it.
+        """
+        requests = list(self.get_requests_by_status(RequestStatus.PENDING))
+        # prioritize the oldest requests at the highest priority value
+        requests.sort(key=lambda x: x["submitted_at"])
+        requests.sort(key=lambda x: x["priority"], reverse=True)
+        for request in requests:
+            self._handle_requested_resources(request)
+
+    def _handle_requested_resources(self, request_entry: dict[str, Any]):
+        try:
+            resource_request = request_entry["request"]
+            task_id = request_entry["task_id"]
+
+            task_status = self.task_view.get_status(task_id=task_id)
+            if task_status != TaskStatus.REQUESTING_RESOURCES:
+                # this implies the Task has been cancelled or errored somewhere else in the chain -- we should
+                # not allocate any resources to the broken Task.
+                self.update_request_status(
+                    request_id=resource_request["_id"],
+                    status=RequestStatus.CANCELED,
+                )
+                return
+
+            devices = self.device_view.request_devices(
+                task_id=task_id,
+                device_names_str=[
+                    entry["device"]["content"]
+                    for entry in resource_request
+                    if entry["device"]["identifier"] == "name"
+                ],
+                device_types_str=[
+                    entry["device"]["content"]
+                    for entry in resource_request
+                    if entry["device"]["identifier"] == "type"
+                ],
+            )
+            # some devices are not available now
+            # the request cannot be fulfilled
+            if devices is None:
```

Comment:
is not it better to be on hold when the request cannot be fulfilled? what is going to happen in this case?

Reply:
If the request cannot be fulfilled (e.g. `if devices is None: ...`), it will check other requests to see if any can be fulfilled. If not, we will just want and recheck again very soon.

Basically, it is what [this](https://github.com/CederGroupHub/alabos/pull/65/files/9b31dbbf8ab7575ebab245f7daa7f45aeb03c0c8#diff-c92448277a32d2fafe81a993dd72b019fd67b2a0ab4087660f32eaa10a4b3d07R53) function does.